File: 백준/Gold/10423. 전기가 부족해/전기가 부족해.java
import java.util.*;
import java.io.*;

public class Main {
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		K = Integer.parseInt(st.nextToken());
		
		uf = new int[N+1];
		for(int i = 0; i <=N; i++) {
			uf[i] = i;
		}
		installed = new boolean[N+1];
		st = new StringTokenizer(br.readLine());
		for(int i = 0; i < K; i++) {
			int idx = Integer.parseInt(st.nextToken());
			installed[idx] = true;
		}
		
		edges = new Edge[M];
		for(int i = 0; i <M;i++) {
			st = new StringTokenizer(br.readLine());
			int u = Integer.parseInt(st.nextToken());
			int v = Integer.parseInt(st.nextToken());
			int cost = Integer.parseInt(st.nextToken());
			edges[i] = new Edge(u,v,cost);
		}
		
		Arrays.sort(edges);
		
		int ans = 0;
		for(Edge edge : edges) {
			int A = find(edge.u);
			int B = find(edge.v);
			
			if(A == B)
				continue;
			
			if(installed[A] && installed[B])
				continue;
			
			if(!installed[A])
				uf[A] = B;
			else
				uf[B] = A;
			
			ans += edge.cost;
			
				
					
		}
		
		System.out.println(ans);
	}
	
	static int find(int x) {
		if(x == uf[x])
			return x;
		return uf[x] = find(uf[x]);
	}
	
	static StringTokenizer st;
	static boolean installed[];
	static int N,M,K;
	static int uf[];
	static Edge edges[];
	static class Edge implements Comparable<Edge>{
		int u,v,cost;
		public Edge(int u,int v,int cost) {
			this.u = u;
			this.v = v;
			this.cost = cost;
		}
		
		public int compareTo(Edge o) {
			return cost - o.cost;
		}
	}
}

File: 백준/Gold/10423. 전기가 부족해/전기가 부족해.py
class Edge:
    def __init__(self, a,b,cost):
        self.a = a
        self.b = b
        self.cost = cost

    def __lt__(self, other):
        return self.cost < other.cost

    def __repr__(self):
        return f"[ ({self.a},{self.b}) : {self.cost} ]"

def find(x):
    if x == uf[x]:
        return x

    uf[x] = find(uf[x])
    return uf[x]

if __name__ == "__main__":
    N,M,K = map(int,input().split())
    edges = []
    installed = set(map(int,input().split()))
    uf = [ i for i in range(N+1)]

    for _ in range(M):
        u,v,w = map(int,input().split())
        edges.append(Edge(u,v,w))

    edges.sort()

    ans = 0

    for edge in edges:
        # print("------")
        # print(edge)
        a = edge.a
        b = edge.b
        cost = edge.cost

        A = find(a)
        B = find(b)

        if A == B:
            continue

        if (A in installed) and (B in installed):
            continue

        if A not in installed:
            uf[A] = B
            ans += cost
        else:
            uf[B] = A
            ans += cost
    # print(uf)
    print(ans)